cwltool/provenance.py

```diff
@@ -0,0 +1,676 @@
+from __future__ import absolute_import
+import io
+import json
+import os
+import os.path
+import posixpath
+import shutil
+import tempfile
+import itertools
+import logging
+import hashlib
+from shutil import copyfile
+import io
+import time
+import copy
+import datetime
+import prov.model as prov
+from prov.model import Namespace, PROV
+from pathlib2 import Path
+# Disabled due to excessive transitive dependencies
+#from networkx.drawing.nx_agraph import graphviz_layout
+#from networkx.drawing.nx_pydot import write_dot
+from .errors import WorkflowException
+import prov.graph as graph
+import uuid
+import graphviz
+import networkx as nx
+import ruamel.yaml as yaml
+import warnings
+from typing import Any, Dict
+from subprocess import check_call
+from schema_salad.sourceline import SourceLine
+from .process import shortname
+
+warnings.simplefilter('ignore', yaml.error.UnsafeLoaderWarning)
+relativised_input_object={}  # type: Dict[str, Any]
+_logger = logging.getLogger("cwltool")
+
+# RO folders
+METADATA = "metadata"
+DATA = "data"
+WORKFLOW = "workflow"
+SNAPSHOT = "snapshot"
+# sub-folders
+MAIN = os.path.join(WORKFLOW, "main")
+PROVENANCE = os.path.join(METADATA, "provenance")
+WFDESC=Namespace("wfdesc", 'http://purl.org/wf4ever/wfdesc#')
+WFPROV=Namespace("wfprov", 'http://purl.org/wf4ever/wfprov#')
+
+# BagIt and YAML always use UTF-8
+ENCODING="UTF-8"
+
+class ProvenanceException(BaseException):
+    pass
+
+
+# sha1, compatible with the File type's "checksum" field
+# e.g. "checksum" = "sha1$47a013e660d408619d894b20806b1d5086aab03b"
+# See ./cwltool/schemas/v1.0/Process.yml
+hashmethod = hashlib.sha1
+
+class ResearchObject():
+    def __init__(self, tmpPrefix="tmp"):
+        self.tmpPrefix = tmpPrefix
+        # type: (...) -> None
```

Comment:
this type annotation belongs one line up :-)

Reply:
Fixed :) How can we test if more type annotations are needed?